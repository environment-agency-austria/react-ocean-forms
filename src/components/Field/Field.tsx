/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { getDeepValue } from '../../utils';
import { IFormContext, TFieldValues } from '../FormContext';
import { IValidationArgs, IValidationState, withValidation } from '../withValidation';
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldProps, IValueMeta, TFieldValue } from './Field.types';

interface IContextMeta extends IValueMeta {
  defaultValue?: TFieldValue;
  externalValue?: TFieldValue;
}

interface IFieldState {
  touched: boolean;
  dirty: boolean;
  value: TFieldValue;
  contextMeta: IContextMeta;
}

interface IMetaChanges {
  defaultValue: boolean;
  externalValue: boolean;
  meta: boolean;
}

/**
 * Wrapper for input fields managed by
 * the form component
 */
export class BaseField extends React.Component<IFieldProps, IFieldState> {
  public static displayName: string = 'Field';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    onChange: (): void => undefined,
    onBlur: (): void => undefined,
    getDisplayValue: (value: TFieldValue): TFieldValue => value,
    getSubmitValue: (value: TFieldValue): TFieldValue => value,
  };

  constructor(props: IFieldProps) {
    super(props);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    context.registerField(
      fullName,
      {
        label,

        updateValidation,
        validate: this.validate,
        reset: this.reset,
        getValue: this.getValue,
      },
    );

    this.state = {
      touched: false,
      dirty: false,
      value: BaseField.callGetDisplayValue(props, ''),
      contextMeta: {
        disabled: false,
        plaintext: false,
      },
    };
  }

  /**
   * Changes the field value based on changes in Form.defaultValues,
   * Form.values, Field.defaultValue or Field.value. The Field props
   * always override the Form props. Changes in defaultValue will only
   * update the field value, if there is no values or value prop present
   * and if the field hasn't been touched.
   */
  public static getDerivedStateFromProps(nextProps: IFieldProps, prevState: IFieldState): IFieldState | null {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = nextProps;

    // Get the default value from Field.defaultValue or Form.defaultValues
    const defaultValue = BaseField.getDefaultValue(nextProps);
    // Get the "external" value from Field.value or Form.values
    const externalValue = BaseField.getExternalValue(nextProps);

    // Get changes in default or external value and meta information (plaintext, disabled)
    // Meta information must trigger a state change as well, because they are passed to
    // Field.getDisplayValue and could result in a different return value of that function
    const changes = BaseField.getPropChanges(nextProps, prevState, defaultValue, externalValue);
    // Get the new value based on the information above, or undefined if we don't need any
    // value changes.
    const propValue = BaseField.getPropValue(defaultValue, externalValue, changes);

    // Update the Field state if needed, also remember the current prop values so we can
    // detect changes in the next call of getDerivedStateFromProps.
    if (propValue !== undefined) {
      return ({
        value: BaseField.callGetDisplayValue(nextProps, propValue),
        touched: false,
        dirty: false,
        contextMeta: {
          defaultValue,
          externalValue,
          disabled: newDisabled,
          plaintext: newPlaintext,
        },
      });
    }

    return null;
  }

  /**
   * Correctly calls the getDisplayValue callback
   * @param props Field props
   * @param value Field value
   */
  private static callGetDisplayValue(props: IFieldProps, value: TFieldValue | undefined): TFieldValue {
    const { getDisplayValue, context } = props;

    return getDisplayValue(
      value || '',
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Calculates the new state value based on the parameters - it will return
   * the externalValue if it exists and has changes, or the default value
   * otherwise. If no state update is needed it will return undefined. Meta
   * changes will trigger a state change as well.
   * @param defaultValue Default value
   * @param externalValue External value
   * @param changes Changes object
   */
  private static getPropValue(
    defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined, changes: IMetaChanges,
  ): TFieldValue | undefined {
    const hasExternalValue = externalValue !== undefined;

    if (hasExternalValue && (changes.externalValue || changes.meta)) {
      return externalValue;
    }

    if (!hasExternalValue && (changes.defaultValue || changes.meta)) {
      return defaultValue;
    }

    return undefined;
  }

  /**
   * Detects changes in the defaultValue, the externalValue and in
   * context.disabled or context.plaintext (merged as meta prop)
   * @param props Field props
   * @param state Field state
   * @param defaultValue Default value
   * @param externalValue External value
   */
  private static getPropChanges(
    props: IFieldProps, state: IFieldState, defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined,
  ): IMetaChanges {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = props;

    const {
      contextMeta: {
        defaultValue: oldDefaultValue,
        externalValue: oldExternalValue,
        disabled: oldDisabled,
        plaintext: oldPlaintext,
      },
      touched,
    } = state;

    return {
      defaultValue: !touched && defaultValue !== oldDefaultValue,
      externalValue: externalValue !== oldExternalValue,
      meta: newDisabled !== oldDisabled || newPlaintext !== oldPlaintext,
    };
  }

  /**
   * Returns the local value if existing, otherwise tries to
   * extract the correct context value based on the fullName.
   * @param localValue Local value
   * @param contextValue Context value
   * @param fullName Field.fullName
   */
  private static getLocalOverridenValue(
    localValue: TFieldValue | undefined, contextValue: TFieldValues | undefined, fullName: string,
  ): TFieldValue | undefined {
    return localValue || getDeepValue(fullName, contextValue);
  }

  /**
   * Returns either the Field.value or the correct value from
   * Form.values
   * @param props Field props
   */
  private static getExternalValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { values }, value } = props;

    return BaseField.getLocalOverridenValue(value, values, fullName);
  }

  /**
   * Returns either the Field.defaultValue or the correct default value
   * from Form.defaultValues
   * @param props Field props
   */
  private static getDefaultValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { defaultValues }, defaultValue } = props;

    return BaseField.getLocalOverridenValue(defaultValue, defaultValues, fullName);
  }

  /**
   * Returns a meta object for the value lifecycle hooks
   * @param context Form context
   */
  private static getValueMeta(context: IFormContext): IValueMeta {
    return {
      disabled: context.disabled,
      plaintext: context.plaintext,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Returns the current field value
   */
  private getValue = (): TFieldValue => {
    const { value } = this.state;
    const { getSubmitValue, context } = this.props;

    return getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  private getAsyncValidateOnChangeSetting(): boolean {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === undefined ? contextChange : propChange;
  }

  /**
   * Checks if the Field is inside a valid form context
   * and throws an user friendly error if not
   */
  private checkFormContext(): void {
    const { context, fullName } = this.props;
    if (!context || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Resets the field to its default state
   */
  private reset = (): void => {
    const { contextMeta: { externalValue, defaultValue } } = this.state;
    const { validation, onChange } = this.props;

    const value = BaseField.callGetDisplayValue(
      this.props,
      externalValue || defaultValue,
    );

    this.setState({
      value,
      touched: false,
      dirty: false,
    });

    validation.reset();
    onChange(value);
  }

  /**
   * Validates the field
   * @param args Validation arguments
   */
  private validate = async (args?: Partial<IValidationArgs>): Promise<IValidationState> => {
    const { value } = this.state;
    const { validation: { validate }, getSubmitValue, context } = this.props;

    return validate(
      getSubmitValue(
        value,
        BaseField.getValueMeta(context),
      ),
      args,
    );
  }

  /**
   * Handles the change event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   * @param event Event object
   */
  private handleFieldChanged = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { value } = event.target;
    const {
      fullName,
      context,
      validation: { validate },
      getSubmitValue,
      onChange,
    } = this.props;

    this.setState({
      value,
      dirty: true,
      touched: true,
    });

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    validate(
      submitValue,
      { checkAsync: asyncValidateOnChange },
    );

    context.notifyFieldEvent(fullName, 'change', submitValue);
    onChange(submitValue);
  }

  /**
   * Handles the blur event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   */
  private handleFieldBlurred = (): void => {
    const {
      fullName,
      validation: { validate },
      context,
      onBlur,
      getSubmitValue,
    } = this.props;
    const { value, dirty } = this.state;

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    if (dirty && !asyncValidateOnChange) { validate(submitValue); }
    context.notifyFieldEvent(fullName, 'blur');
    onBlur();
  }

  /**
   * Creates the properties that can be directly
   * mapped to the input component (e.g. html input)
   */
  private createFieldProps(): IFieldComponentFieldProps {
    const { context: { disabled }, fullName } = this.props;
    const { value } = this.state;

    return {
      value,
      disabled,
      id: fullName,
      name: fullName,
      onChange: this.handleFieldChanged,
      onBlur: this.handleFieldBlurred,
    };
  }

  /**
   * Creates the meta properties with meta information
   * used by the input component
   */
  private createMetaProps(): IFieldComponentMeta {
    const {
      context: {
        stringFormatter,
        plaintext,
      },
      validation: {
        valid,
        error,
        isValidating,
      },
    } = this.props;

    const { touched } = this.state;

    return {
      valid,
      error,
      isValidating,
      touched,
      stringFormatter,
      plaintext,
    };
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      // tslint:disable-next-line:naming-convention
      component: InputComponent,
      context,
      fullName,
      validation,
      onChange,
      onBlur,
      defaultValue,
      value,
      name,
      asyncValidateOnChange,
      getDisplayValue,
      getSubmitValue,
      ...attributes
    } = this.props;

    const field = this.createFieldProps();
    const meta = this.createMetaProps();

    return (
      <InputComponent
        {...attributes}
        field={field}
        meta={meta}
      />
    );
  }
}

export const Field = withValidation(BaseField);
