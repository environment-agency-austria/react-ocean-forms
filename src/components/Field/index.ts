export * from './Field';
export * from './Field.types';
