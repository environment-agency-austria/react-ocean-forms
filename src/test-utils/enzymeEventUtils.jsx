/**
 * Defines some often used keycodes for
 * event testing
 */
export const KEYCODE = {
  BACKSPACE: 8,
  TAB: 9,
  ENTER: 13,
  ESCAPE: 27,
  DOWN_ARROW: 40,
};

/**
 * Generates a mocked event to be used with
 * enzymes simulate in situations where the
 * tested code wants to call preventDefault
 * @param {object} props Additional event props
 */
export const mockEvent = props => ({
  preventDefault: () => {},
  ...props,
});

export default mockEvent;
