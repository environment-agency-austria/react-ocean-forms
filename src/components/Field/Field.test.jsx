import React from 'react';
import { shallow } from 'enzyme';

import { createMockFormContext, createMockValidation } from '../../test-utils/enzymeFormContext';
import { BaseField } from './Field';

describe('<Field />', () => {
  const MOCK_NAME = 'unitField';
  const MOCK_LABEL = 'Unit field';

  const setup = ({
    props = undefined,
    contextOverrides = undefined,
  } = {}) => {
    let fieldState = null;
    const registerCallback = (name, state) => { fieldState = state; };

    const formContext = {
      ...createMockFormContext(registerCallback),
      ...contextOverrides,
    };
    const validation = createMockValidation();

    const TestComponent = () => (<div id="test-component" />);

    const wrapper = shallow((
      <BaseField
        name={MOCK_NAME}
        fullName={MOCK_NAME}
        label={MOCK_LABEL}
        component={TestComponent}
        context={formContext}
        validation={validation}
        {...props}
      />
    ));

    return {
      formContext,
      fieldState,
      validation,
      wrapper,
    };
  };

  const getTestComponent = wrapper => wrapper.find('TestComponent');
  const getValue = wrapper => getTestComponent(wrapper).prop('field').value;
  const assertValue = (wrapper, value) => expect(getValue(wrapper)).toBe(value);
  const simulateChange = (wrapper, value) => getTestComponent(wrapper).prop('field').onChange({
    target: {
      name: MOCK_NAME,
      value,
    },
  });
  const simulateBlur = wrapper => getTestComponent(wrapper).prop('field').onBlur();

  describe('Render', () => {
    it('should render without crashing', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Form registration', () => {
    let formContext;
    let validation;
    let wrapper;

    beforeAll(() => {
      ({ formContext, validation, wrapper } = setup());
    });

    it('should register itself in the form context', () => {
      expect(formContext.registerField).toHaveBeenCalledWith(
        MOCK_NAME,
        {
          label: MOCK_LABEL,
          validate: expect.any(Function),
          reset: expect.any(Function),
          getValue: expect.any(Function),
          updateValidation: validation.update,
        },
      );
    });

    it('should unregister itself on unmount', () => {
      wrapper.unmount();
      expect(formContext.unregisterField).toHaveBeenCalledWith(MOCK_NAME);
    });
  });

  describe('Invalid form context', () => {
    const MOCK_ERROR_STRING = `Could not find a form context for field "${MOCK_NAME}". `
                            + 'Fields can only be used inside a Form tag.';

    it('should throw an error if there is no form context', () => {
      expect(() => setup({ props: { context: undefined } })).toThrowError(MOCK_ERROR_STRING);
    });

    it('should throw an error if the form context is invalid', () => {
      expect(() => setup({ props: { context: { foo: 'bar' } } })).toThrowError(MOCK_ERROR_STRING);
    });
  });

  describe('Default value handling', () => {
    let wrapper;

    afterEach(() => wrapper.unmount());

    it('should have an empty string as the default value', () => {
      ({ wrapper } = setup());
      assertValue(wrapper, '');
    });

    it('should use the default value from Form.defaultValues if existing', () => {
      const MOCK_DEFAULTVALUE = 'mock-default-value';

      ({ wrapper } = setup({
        contextOverrides: {
          defaultValues: { [MOCK_NAME]: MOCK_DEFAULTVALUE },
        },
      }));

      assertValue(wrapper, MOCK_DEFAULTVALUE);
    });

    it('should use the Field.defaultValue if existing', () => {
      const MOCK_DEFAULTVALUE = 'mock-field-value';

      ({ wrapper } = setup({
        props: {
          defaultValue: MOCK_DEFAULTVALUE,
        },
      }));

      assertValue(wrapper, MOCK_DEFAULTVALUE);
    });

    it('should prefer the Field.defaultValue over the Form.defaultValues', () => {
      const MOCK_DEFAULTVALUE_FIELD = 'mock-field-value';
      const MOCK_DEFAULTVALUE_FORM = 'mock-form-value';

      ({ wrapper } = setup({
        props: {
          defaultValue: MOCK_DEFAULTVALUE_FIELD,
        },
        contextOverrides: {
          defaultValues: { [MOCK_NAME]: MOCK_DEFAULTVALUE_FORM },
        },
      }));

      assertValue(wrapper, MOCK_DEFAULTVALUE_FIELD);
    });

    it('should not use the defaultValue if the Field is touched', () => {
      let formContext;

      const MOCK_DEFAULT_VALUE = 'mock-field-value';
      const MOCK_CHANGE_VALUE = 'mock-change-value';

      ({ wrapper, formContext } = setup()); // eslint-disable-line

      // Recreate a field value change
      simulateChange(wrapper, MOCK_CHANGE_VALUE);

      // Set new defaultProps through Form.defaultValues
      formContext.defaultValues = { [MOCK_NAME]: MOCK_DEFAULT_VALUE };
      wrapper.setProps({ context: formContext });
      assertValue(wrapper, MOCK_CHANGE_VALUE);

      // Set new defaultProps through Field.defaultValue
      wrapper.setProps({ defaultValue: MOCK_DEFAULT_VALUE });
      assertValue(wrapper, MOCK_CHANGE_VALUE);
    });
  });

  describe('Prop value handling', () => {
    let wrapper;

    afterEach(() => wrapper.unmount());

    it('should use the value from Form.values if existing', () => {
      const MOCK_VALUE = 'mock-value';

      ({ wrapper } = setup({
        contextOverrides: {
          values: { [MOCK_NAME]: MOCK_VALUE },
        },
      }));

      assertValue(wrapper, MOCK_VALUE);
    });

    it('should use the Field.value if existing', () => {
      const MOCK_VALUE = 'mock-field-value';

      ({ wrapper } = setup({
        props: {
          value: MOCK_VALUE,
        },
      }));

      assertValue(wrapper, MOCK_VALUE);
    });

    it('should prefer the Field.value over the Form.values', () => {
      const MOCK_VALUE_FIELD = 'mock-field-value';
      const MOCK_VALUE_FORM = 'mock-form-value';

      ({ wrapper } = setup({
        props: {
          value: MOCK_VALUE_FIELD,
        },
        contextOverrides: {
          values: { [MOCK_NAME]: MOCK_VALUE_FORM },
        },
      }));

      assertValue(wrapper, MOCK_VALUE_FIELD);
    });

    it('Field.value should override the default values', () => {
      const MOCK_VALUE = 'mock-field-value';
      const MOCK_DEFAULT_VALUE = 'mock-default-value';

      ({ wrapper } = setup({
        props: {
          value: MOCK_VALUE,
          defaultValue: MOCK_DEFAULT_VALUE,
        },
        contextOverrides: {
          defaultValues: { [MOCK_NAME]: MOCK_DEFAULT_VALUE },
        },
      }));

      assertValue(wrapper, MOCK_VALUE);
    });

    it('Form.values should override the default values', () => {
      const MOCK_VALUE = 'mock-field-value';
      const MOCK_DEFAULT_VALUE = 'mock-default-value';

      ({ wrapper } = setup({
        props: {
          defaultValue: MOCK_DEFAULT_VALUE,
        },
        contextOverrides: {
          values: { [MOCK_NAME]: MOCK_VALUE },
          defaultValues: { [MOCK_NAME]: MOCK_DEFAULT_VALUE },
        },
      }));

      assertValue(wrapper, MOCK_VALUE);
    });

    it('should use the changed value even if the Field is touched', () => {
      let formContext;

      let MOCK_VALUE = 'mock-field-value';
      const MOCK_CHANGE_VALUE = 'mock-change-value';

      const updateValue = () => {
        // Recreate a field value change
        simulateChange(wrapper, MOCK_CHANGE_VALUE);
        assertValue(wrapper, MOCK_CHANGE_VALUE);
      };

      ({ wrapper, formContext } = setup()); // eslint-disable-line

      // Mock user input
      updateValue();

      // Set new value through Form.values
      formContext.values = { [MOCK_NAME]: MOCK_VALUE };
      wrapper.setProps({ context: formContext });
      assertValue(wrapper, MOCK_VALUE);

      // Mock user input
      updateValue();

      // Set new props through Field.value
      MOCK_VALUE = 'mock-new-field-value';
      wrapper.setProps({ value: MOCK_VALUE });
      assertValue(wrapper, MOCK_VALUE);
    });
  });

  describe('onChange handling', () => {
    const MOCK_VALUE = 'mock-change-value';

    let wrapper;
    let validation;
    let formContext;

    const setupOnChange = (props, contextOverrides) => {
      ({ wrapper, validation, formContext } = setup({
        props,
        contextOverrides,
      }));
      simulateChange(wrapper, MOCK_VALUE);
    };

    afterEach(() => {
      wrapper.unmount();
      validation = undefined;
      formContext = undefined;
    });

    it('should remember the changed value', () => {
      setupOnChange();
      assertValue(wrapper, MOCK_VALUE);
    });

    it('should call the Field.getSubmitValue callback', () => {
      const MOCK_GET_SUBMIT_VALUE = jest.fn().mockImplementation(value => value);
      setupOnChange({ getSubmitValue: MOCK_GET_SUBMIT_VALUE });

      expect(MOCK_GET_SUBMIT_VALUE).toHaveBeenCalledWith(
        MOCK_VALUE,
        { disabled: false, plaintext: false },
      );
    });

    it('should call the validate function', () => {
      setupOnChange();
      expect(validation.validate).toHaveBeenCalledWith(
        MOCK_VALUE,
        { checkAsync: false },
      );
    });

    it('should notify the form context', () => {
      setupOnChange();
      expect(formContext.notifyFieldEvent).toHaveBeenCalledWith(
        MOCK_NAME,
        'change',
        MOCK_VALUE,
      );
    });

    it('should call the Field.onChange handler', () => {
      const MOCK_ON_CHANGE = jest.fn();
      setupOnChange({ onChange: MOCK_ON_CHANGE });
      expect(MOCK_ON_CHANGE).toHaveBeenCalledWith(MOCK_VALUE);
    });

    it('should respect the Form.asyncValidateOnChange configuration', () => {
      const MOCK_CHECK_ASYNC = true;
      setupOnChange(null, { asyncValidateOnChange: MOCK_CHECK_ASYNC });
      expect(validation.validate).toHaveBeenCalledWith(
        MOCK_VALUE,
        { checkAsync: MOCK_CHECK_ASYNC },
      );
    });

    it('should respect the Field.asyncValidateOnChange configuration', () => {
      const MOCK_CHECK_ASYNC = true;
      setupOnChange({ asyncValidateOnChange: MOCK_CHECK_ASYNC });

      expect(validation.validate).toHaveBeenCalledWith(
        MOCK_VALUE,
        { checkAsync: MOCK_CHECK_ASYNC },
      );
    });
  });

  describe('onBlur handling', () => {
    const MOCK_VALUE = 'mock-value';

    let wrapper;
    let validation;
    let formContext;

    const setupLocal = (props, contextOverrides) => {
      ({ wrapper, validation, formContext } = setup({
        props: { ...props, value: MOCK_VALUE },
        contextOverrides,
      }));
    };

    const setupOnBlur = (props, contextOverrides) => {
      setupLocal(props, contextOverrides);
      simulateBlur(wrapper);
    };

    afterEach(() => {
      wrapper.unmount();
      validation = undefined;
      formContext = undefined;
    });

    it('should call the Field.getSubmitValue function', () => {
      const MOCK_GET_SUBMIT_VALUE = jest.fn().mockImplementation(value => value);
      setupOnBlur({ getSubmitValue: MOCK_GET_SUBMIT_VALUE });

      expect(MOCK_GET_SUBMIT_VALUE).toHaveBeenCalledWith(
        MOCK_VALUE,
        { disabled: false, plaintext: false },
      );
    });

    it('should call the validate function', () => {
      const MOCK_CHANGE_VALUE = 'foo';
      setupLocal();

      simulateChange(wrapper, MOCK_CHANGE_VALUE);
      validation.validate.mockClear();

      simulateBlur(wrapper);
      expect(validation.validate).toHaveBeenCalledWith(MOCK_CHANGE_VALUE);
    });

    it('should not call the validate function if the field is not dirty', () => {
      setupOnBlur();
      expect(validation.validate).not.toHaveBeenCalledWith(MOCK_VALUE);
    });

    it('should notify the form context', () => {
      setupOnBlur();
      expect(formContext.notifyFieldEvent).toHaveBeenCalledWith(
        MOCK_NAME,
        'blur',
      );
    });

    it('should call the Field.onBlur handler', () => {
      const MOCK_ON_BLUR = jest.fn();
      setupOnBlur({ onBlur: MOCK_ON_BLUR });
      expect(MOCK_ON_BLUR).toHaveBeenCalled();
    });
  });

  describe('getDisplayValue handling', () => {
    const MOCK_DISPLAY_VALUE = 'mock-display-value';
    const MOCK_VALUE = 'mock-value';

    const setupWithDisplayName = () => {
      const MOCK_GET_DISPLAY_VALUE = jest.fn().mockReturnValue(MOCK_DISPLAY_VALUE);
      return {
        ...setup({ props: { getDisplayValue: MOCK_GET_DISPLAY_VALUE, value: MOCK_VALUE } }),
        getDisplayValue: MOCK_GET_DISPLAY_VALUE,
      };
    };

    it('should call Field.getDisplayValue on first render', () => {
      const { wrapper, getDisplayValue } = setupWithDisplayName();
      assertValue(wrapper, MOCK_DISPLAY_VALUE);
      expect(getDisplayValue).toHaveBeenCalledWith(
        '',
        { disabled: false, plaintext: false },
      );
    });

    it('should call getDisplayValue whenever the Context.disabled state changes', () => {
      const { wrapper, formContext, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      formContext.disabled = true;
      wrapper.setProps({ context: formContext });

      expect(getDisplayValue).toHaveBeenCalledWith(
        MOCK_VALUE,
        { disabled: true, plaintext: false },
      );
    });

    it('should call getDisplayValue whenever the Context.plaintext state changes', () => {
      const { wrapper, formContext, getDisplayValue } = setupWithDisplayName();
      getDisplayValue.mockClear();

      formContext.plaintext = true;
      wrapper.setProps({ context: formContext });

      expect(getDisplayValue).toHaveBeenCalledWith(
        MOCK_VALUE,
        { disabled: false, plaintext: true },
      );
    });
  });

  describe('Form context callbacks', () => {
    describe('Context.getValue', () => {
      it('should return the correct value', () => {
        const MOCK_VALUE = 'field-value';
        const { fieldState } = setup({ props: { value: MOCK_VALUE } });
        expect(fieldState.getValue()).toBe(MOCK_VALUE);
      });

      it('should pass the return value through getSubmitValue', () => {
        const MOCK_VALUE = 'field-value';
        const MOCK_SUBMIT_VALUE = 'field-submit-value';
        const MOCK_GET_SUBMIT_VALUE = jest.fn().mockReturnValue(MOCK_SUBMIT_VALUE);

        const { fieldState } = setup({
          props: {
            value: MOCK_VALUE,
            getSubmitValue: MOCK_GET_SUBMIT_VALUE,
          },
        });

        expect(fieldState.getValue()).toBe(MOCK_SUBMIT_VALUE);
        expect(MOCK_GET_SUBMIT_VALUE).toHaveBeenCalledWith(
          MOCK_VALUE,
          { disabled: false, plaintext: false },
        );
      });
    });

    describe('Context.validate', () => {
      it('should correctly pass the validate method', () => {
        const MOCK_VALIDATE_ARGS = { foo: 'bar' };
        const MOCK_VALUE = 'field-value';

        const { fieldState, validation } = setup({ props: { value: MOCK_VALUE } });
        fieldState.validate(MOCK_VALIDATE_ARGS);

        expect(validation.validate).toHaveBeenLastCalledWith(
          MOCK_VALUE,
          MOCK_VALIDATE_ARGS,
        );
      });

      it('should pass the validated value through getSubmitValue', () => {
        const MOCK_VALUE = 'field-value';
        const MOCK_SUBMIT_VALUE = 'field-submit-value';
        const MOCK_GET_SUBMIT_VALUE = jest.fn().mockReturnValue(MOCK_SUBMIT_VALUE);

        const { fieldState } = setup({
          props: {
            value: MOCK_VALUE,
            getSubmitValue: MOCK_GET_SUBMIT_VALUE,
          },
        });

        fieldState.validate();
        expect(MOCK_GET_SUBMIT_VALUE).toHaveBeenCalledWith(
          MOCK_VALUE,
          { disabled: false, plaintext: false },
        );
      });
    });

    describe('Context.reset', () => {
      it('should reset its validation state', () => {
        const { fieldState, validation } = setup();

        fieldState.reset();
        expect(validation.reset).toHaveBeenCalled();
      });


      const cases = [
        ['Field.defaultValue', 'defaultValue', undefined],
        ['Field.value', 'value', undefined],
        ['Form.defaultValues', undefined, 'defaultValues'],
        ['Form.values', undefined, 'values'],
      ];

      describe.each(cases)('Reset to %s', (name, prop, context) => {
        const MOCK_DEFAULT_VALUE = 'field-default-value';
        const MOCK_CHANGE_VALUE = 'field-change-value';

        const props = prop ? { [prop]: MOCK_DEFAULT_VALUE } : undefined;
        const contextOverrides = context
          ? { [context]: { [MOCK_NAME]: MOCK_DEFAULT_VALUE } }
          : undefined;

        it('should correctly reset to its defaultValue', () => {
          const { wrapper, fieldState } = setup({ props, contextOverrides });

          simulateChange(wrapper, MOCK_CHANGE_VALUE);
          fieldState.reset();

          assertValue(wrapper, MOCK_DEFAULT_VALUE);
        });

        it('should call the Field.getDisplayValue function', () => {
          const MOCK_DISPLAY_VALUE = 'mock-display-value';
          const MOCK_GET_DISPLAY_VALUE = jest.fn().mockReturnValue(MOCK_DISPLAY_VALUE);

          const { wrapper, fieldState } = setup({
            props: {
              ...props,
              getDisplayValue: MOCK_GET_DISPLAY_VALUE,
            },
            contextOverrides,
          });

          simulateChange(wrapper, MOCK_CHANGE_VALUE);
          fieldState.reset();

          assertValue(wrapper, MOCK_DISPLAY_VALUE);
          expect(MOCK_GET_DISPLAY_VALUE).toHaveBeenCalledWith(
            MOCK_DEFAULT_VALUE,
            { disabled: false, plaintext: false },
          );
        });

        it('should call the onChange handler', () => {
          const MOCK_ON_CHANGE = jest.fn();
          const { fieldState } = setup({
            props: {
              onChange: MOCK_ON_CHANGE,
              ...props,
            },
            contextOverrides,
          });

          fieldState.reset();
          expect(MOCK_ON_CHANGE).toHaveBeenCalledWith(MOCK_DEFAULT_VALUE);
        });
      });
    });
  });
});
