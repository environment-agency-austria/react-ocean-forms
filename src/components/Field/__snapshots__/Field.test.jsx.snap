// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> Render should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Field
    component={[Function]}
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction] {
          "calls": Array [
            Array [
              "unitField",
              Object {
                "getValue": [Function],
                "label": "Unit field",
                "reset": [Function],
                "updateValidation": [MockFunction],
                "validate": [Function],
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    fullName="unitField"
    getDisplayValue={[Function]}
    getSubmitValue={[Function]}
    label="Unit field"
    name="unitField"
    onBlur={[Function]}
    onChange={[Function]}
    validation={
      Object {
        "error": null,
        "isValidating": false,
        "reset": [MockFunction],
        "update": [MockFunction],
        "valid": true,
        "validate": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "field": Object {
        "disabled": false,
        "id": "unitField",
        "name": "unitField",
        "onBlur": [Function],
        "onChange": [Function],
        "value": "",
      },
      "label": "Unit field",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "plaintext": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "field": Object {
          "disabled": false,
          "id": "unitField",
          "name": "unitField",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "",
        },
        "label": "Unit field",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "plaintext": false,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
